enhancements/microshift/microshift-tls-cipher.md

```diff
@@ -0,0 +1,233 @@
+MicroShift's exposed endpoints (such as the default router and the API server)
+are using TLS. In order to secure the API server only TLS 1.2 or above is
+allowed.
+
+Cipher suites are key to the security of TLS handshakes and information
+exchange, and different versions of TLS use different cipher suites. Some TLS
+versions (like 1.0 and 1.1) have been deprecated in favor of new ones, being
+1.2 the most widespread. As with TLS versions, the cipher suites have also
+evolved and some of them have been compromised and are now unsafe for use.
+For these reasons, the possibility of restricting which cipher suites to use
+becomes important for serving endpoints, as they allow administrators to
+protect their APIs.
+
+TLS 1.3 has had changes in cipher suites to simplify them, meaning the
+available suites has been reduced to only a handful. Most cipher suites from
+TLS 1.2 are incompatible with 1.3. Further details on this topic can be found
+in the proposal section.
+
+MicroShift API server is using TLS 1.2 as the minimum version, supporting also
+1.3. Cluster admins who are using tools with TLS 1.2 or above may need to
+comply with security policies restricting which cipher suites can be used.
+
+### User Stories
+As a MicroShift admin, I want to be able to configure TLS cipher suites for the
+API endpoints.
+
+As a MicroShift admin, I want to be able to configure the minimum TLS version
+for the API endpoints.
+
+### Goals
+* Allow users to configure which TLS version and which TLS ciphers are used in
+  API endpoints of MicroShift.
+
+### Non-Goals
+N/A
+
+## Proposal
+Introduce configuration options for API server cipher suites.
+```yaml
+apiServer:
+  tls:
+    cipherSuites: # Defaults to the suites of the configured minVersion
+    - <cipher suite 1>
+    - ...
+    minVersion: <VersionTLS12|VersionTLS13> # Defaults to 1.2
+```
+The API server will use the configured minimum TLS version (which defaults to
+1.2) and the cipher suites. Note that cipher suites may be unconfigured, in
+which case it will automatically pick up the defaults for the configured
+minimum version.
+
+The default cipher suites for 1.2 are:
+* TLS_AES_128_GCM_SHA256
+* TLS_AES_256_GCM_SHA384
+* TLS_CHACHA20_POLY1305_SHA256
+* TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+* TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+* TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+* TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+* TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
+* TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+
+The default cipher suites for 1.3 are:
+* TLS_AES_128_GCM_SHA256
+* TLS_AES_256_GCM_SHA384
+* TLS_CHACHA20_POLY1305_SHA256
+
+For simplicity, the API server cipher suites will also apply to internal
+control plane components:
+* API server
+* Kubelet
+* Kube controller manager
+* Kube scheduler
+* Etcd
+* Route controller manager
+
+Except for API server the rest of the components are internal to MicroShift
+and are not exposed through an endpoint but require secure connections too.
```

Comment:
I see in code that RCM is using `"0.0.0.0:8445"`. Also, take a look at this:
```
sudo netstat -tulpn | rg "LISTEN.*microshift"
tcp        0      0 127.0.0.1:2379          0.0.0.0:*               LISTEN      454613/microshift-e 
tcp        0      0 127.0.0.1:2381          0.0.0.0:*               LISTEN      454613/microshift-e 
tcp        0      0 127.0.0.1:2380          0.0.0.0:*               LISTEN      454613/microshift-e 
tcp        0      0 127.0.0.1:10248         0.0.0.0:*               LISTEN      454600/microshift   
tcp6       0      0 :::8445                 :::*                    LISTEN      454600/microshift   
tcp6       0      0 :::10257                :::*                    LISTEN      454600/microshift   
tcp6       0      0 :::10259                :::*                    LISTEN      454600/microshift   
tcp6       0      0 :::10250                :::*                    LISTEN      454600/microshift   
tcp6       0      0 :::2112                 :::*                    LISTEN      454600/microshift   
tcp6       0      0 :::6443                 :::*                    LISTEN      454600/microshift 
```

Even though it says tcp6, we're probably exposing them on ipv4 as well (from the host to VM):
```
$ curl -k https://192.168.64.3:10259/healthz
ok
```
So the question is, should we expose anything other than KAS on 0.0.0.0, should it be localhost only? This might be orthogonal to the feature at hand, but may be important as well

Reply:
It is indeed. Some components (like etcd, ports 2379 to 2381, or kubelet with 10248) were restricted to localhost because of this, I guess the others were disregarded by whatever reason. This needs some analysis though.
I rephrased the sentence above.